### PR TITLE
Sync with Facebook upstream 1fc41d1

### PR DESCRIPTION
**Pull request by jmoody**
### Motivation

It has been almost 3 months since we sync'd.

* 0.2.2
* 1fc41d1038f665bcceeddfffd011a1d083f8cb5b
* Fri Feb 10 2017 7:30 CET

### Tests

```
$ cd calabash/FBSimulatorControl
$ git co -t origin/feature/sync-with-facebook-upstream
$ cd calabash/iOSDeviceManager; git co develop; git pull
$ make facebook-frameworks
```

Xcode 8.3 tests will fail at some point because of [Missing Framework libswiftAppKit.dylib in Xcode 8.3 beta 2](https://github.com/facebook/FBSimulatorControl/issues/375).

I did run the unit, integration, and DeviceAgent stack tests against Xcode 8.3 beta 2, but I did not include the results here because I don't think they provide much information.

- [x] DeviceAgent stack
    - [x] Simulators
    - [x] Devices
- [x] `$ make test-unit`
- [x] `$ make test-integration`

### CODE REVIEW

**Comment by jmoody** [comment]
@sapieneptus 

FBFramebuffer no longer has a `framebufferService` property.

I believe you added the `doOnMainAndReturn` as part of the daemonizing of iOSDeviceManager.

It looks like Facebook is expecting that method to be called on the main thread.

Are these two lines important:

```
-      [self.framebufferService registerClient:self onQueue:self.clientQueue];
-      [self.framebufferService resume];
```

### Diff

```
-  return [CalabashUtils doOnMainAndReturn:^id{
-      NSParameterAssert(NSThread.currentThread.isMainThread);
-      NSParameterAssert(self.state == FBSimulatorFramebufferStateNotStarted);
-      self.state = FBSimulatorFramebufferStateStarting;
-      [self.framebufferService registerClient:self onQueue:self.clientQueue];
-      [self.framebufferService resume];
-
-      return self;
-  }];
+  NSParameterAssert(NSThread.currentThread.isMainThread);
+  NSParameterAssert(self.state == FBSimulatorFramebufferStateNotStarted);
+  self.state = FBSimulatorFramebufferStateStarting;
+  return self;
```

**Comment by sapieneptus** [comment]
> It looks like Facebook is expecting that method to be called on the main thread.

I'll need to re-review the specifics, but at the time it was the case that not all methods FB wanted on the main thread needed to be entirely on the main thread. 

Don't let the old dæmonization changes stop you, proceed at will - I'll be re-reviewing the dæmonization work soon enough. A lot has changed since the initial PR.

**Comment by sapieneptus** [comment]
@jmoody there was a subtle but problematic merge conflict that was hidden inside of a block comment. It accidentally exposed a method that's no longer supported. I've made a commit to resolve (`FBDevice.m#UUID`)

**Comment by jmoody** [comment]
@sapieneptus Good catch.

**Comment by jmoody** [comment]
@sapieneptus This is ready for review and merging.  This does not fix execution on Xcode 8.3, but it does give some clues about what might be wrong:  cannot find `libswiftAppKit.dylib`.  Tracking here [TCFW-1013](https://jira.xamarin.com/browse/TCFW-1013).

File: Documentation/SimulatorHID.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+-[DeviceWindow sendKeyboardEvent:]
+```
+
+Objective-C does encode the shape of structures passed to Objective-C Methods in the executable [using the same encoding as is used in `@encode`](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html). You can see this for yourself by dumping the sections of the Simulator's executable:
+
+```
+otool -ov /Applications/Xcode.app/Contents/Developer/Applications/Simulator.app/Contents/MacOS/Simulator | grep -A 1 sendIndigoHID
+```
+
+`class-dump` is capable of reconstructing this ecoding into a `struct`ure definition for a [C Header](../PrivateHeaders/SimulatorApp/Indigio.h).
```
Comment (sapieneptus) [inline Documentation/SimulatorHID.md:30]: typo: 'ecoding' -> 'encoding'

File: Documentation/SimulatorHID.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  contents++;
+  IOHIDEvent nextEvent = __IOHIDEventFromIndigoHIDData(contents, 0x2, 0x0);
+  // Update the original event by appending.
+  IOHIDEventAppendEvent(event, nextEvent);
+}
+// Dispatch the event to the Event Service.
+IOHIDServiceDispatchEvent(service, event);
+```
+
+The `__IOHIDEventFromIndigoHIDData` function is clearly quite important in tranforming `Indigo` messages into `IOHIDEvent` instances. The implementation of this function is also quite helpful in re-constructing an idea of what an Indigo message looks like. Inside the implementation of this transformation function there are number of conditionals that change the kind of `IOHIDEvent` (Keyboard/Touch) depending on the data passed in.
```
Comment (sapieneptus) [inline Documentation/SimulatorHID.md:77]: These guys are very on top of their game. It's nice to see all of this written out.

File: Documentation/SimulatorHID.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+## Tracing with dtrace
+
+Included in the `dtrace` [directory within this directory](dtrace) are a number of [`dtrace`](x-man-page://dtrace) scripts that demonstrate how messages are sent by `Simulator.app` and recieved by `backboardd`.
```
Comment (sapieneptus) [inline Documentation/SimulatorHID.md:81]: @john7doe We might want to take a peek at these

**Comment by sapieneptus** [comment]
@jmoody I don't think I can meaningfully review that many changes. I vote we merge and do regression tests :/

**Comment by jmoody** [comment]
Not ready.  Tested the wrong frameworks.

**Comment by jmoody** [comment]
We use many private methods on `FBiOSDeviceOperator`.  Instead of updating FBiOSDeviceOperator.h in FBSimulatorControl, I decided to make an Objective-C category in `iOSDeviceManager` to expose the private methods.

We define a convenience method on `XCTestBootstrap`.  Again, I used an Objective-C category in `iOSDeviceManager` to expose this method.

I did this because in both cases this sync clobbered (erased) our changes.

Using an Objective-C category on iOSDeviceManager also has the side benefit of disambiguating the owner of the method: us or FBSimulatorControl.

The downside is that I needed to export TestManagerContext.h as a public header.

**Pull request merged**